Realm/ObjectServerTests/RLMObjectServerTests.mm

```diff
@@ -2242,4 +2243,286 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
+                                                               delegateQueue:delegateQueue];
+
+    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (int i = 0; i < 3; i++) {
+            [collection updateOneDocumentWhere:@{@"_id": objectIds[0]}
+                                updateDocument:@{@"breed": @"king charles", @"name": [NSString stringWithFormat:@"fido-%d", i]}
+                                    completion:^(RLMUpdateResult * _Nullable, NSError * error) {
+                XCTAssertNil(error);
+            }];
+
+            [collection updateOneDocumentWhere:@{@"_id": objectIds[1]}
+                                updateDocument:@{@"breed": @"french bulldog", @"name": [NSString stringWithFormat:@"fido-%d", i]}
+                                    completion:^(RLMUpdateResult * _Nullable, NSError * error) {
+                XCTAssertNil(error);
+            }];
+            dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+            if (i == 2) {
+                [changeStream close];
+            }
+        }
+    });
+    [self waitForExpectations:@[expectation] timeout:60.0];
+}
+
+- (void)testWatchWithFilterIds {
+    [self performWatchWithFilterIdsTest:nil];
+}
+
+- (void)testWatchWithFilterIdsAsync {
+    auto asyncQueue = dispatch_queue_create("io.realm.watchQueue", DISPATCH_QUEUE_CONCURRENT);
+    [self performWatchWithFilterIdsTest:asyncQueue];
+}
+
+- (void)performWatchWithFilterIdsTest:(nullable dispatch_queue_t)delegateQueue {
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    RLMMongoClient *client = [self.anonymousUser mongoClientWithServiceName:@"mongodb1"];
+    RLMMongoDatabase *database = [client databaseWithName:@"test_data"];
+    __block RLMMongoCollection *collection = [database collectionWithName:@"Dog"];
+
+    __block NSArray<RLMObjectId *> * objectIds;
+    XCTestExpectation *insertManyExpectation = [self expectationWithDescription:@"should insert one document"];
+    [collection insertManyDocuments:@[
+        @{@"name": @"fido", @"breed": @"cane corso"},
+        @{@"name": @"rex", @"breed": @"tibetan mastiff"},
+        @{@"name": @"john", @"breed": @"tibetan mastiff"}]
+                         completion:^(NSArray<RLMObjectId *> * ids, NSError * error) {
+        XCTAssertEqual((int)ids.count, 3);
+        XCTAssertNil(error);
+        objectIds = ids;
+        [insertManyExpectation fulfill];
+    }];
+    [self waitForExpectations:@[insertManyExpectation] timeout:60.0];
+
+    XCTestExpectation *expectation = [self expectationWithDescription:@"watch collection and receive change event 3 times"];
+
+    RLMWatchTestUtility *testUtility =
+    [[RLMWatchTestUtility alloc] initWithChangeEventCount:3
+                                         matchingObjectId:objectIds[0]
+                                            eventReceived:^(NSError * error) {
+        XCTAssertNil(error);
```

Comment:
It looks like every use of RLMWatchTestUtility has the exact same logic in the eventReceived: and completion: blocks, so that could just be pushed into RLMWatchTestUtility to eliminate some duplication between tests. It'd probably make sense to have the initializer take the XCTest as a parameter and provide the semaphore and expectation as properties.

Reply:
All logic has moved to RLMWatchTestUtility